### DISCUSSION TITLE
How to differentiate between multiple UploadDropZone?

### DISCUSSION THREAD

**Discussion by ardian-c** [rpldy/react-uploady#437] [Q&A] [answered]
Hi all, 

Thank you for this great library. I'm facing an issue or at least I don't see how to add multiple `UploadDropZone` components inside `Uploady` context. 
Currently, I have these two dropzones, one for a logo upload and the other one for document upload (csv). 
I have `useBatchFinishListener` for each component, `LogoUpload` and `DocumentUpload`. 
So, when I upload something say a logo, the listener will show it in both places, which is probably expected behavior for that file listener. 
However, I'm struggling to see how should I differentiate between these dropzones, and how should I use the `UploadDropZone` and listen for the right files in the respective components.
Appreciate any guidance.

**Comment by yoavniran** [comment, marked as answer]
thanks @ardian-c 
This question came up before so I created an [example sandbox](https://codesandbox.io/s/react-uploady-two-fields-with-single-uploady-e059m?file=/src/App.js) that helped (for that use case) - Its based on the UploadButton but should apply just the same.

Let me know if it helps you or not.

**Reply by ardian-c** [reply]
Hi @yoavniran, 

I saw that example, and I thought if there is a way to check in the listener which dropzone element is currently uploading. From the example I'm having problems with this part: 

```
const useTypedBatchMethod = (cb: any): void => {
    useBatchAddListener((batch, options) => {
      if (options?.params?.uploadType === type) {
        cb(batch, options);
      }
    });
};
```
 From what I see in the examples in your repository this only accepts one parameter, the batch not sure about the options part. Maybe I'm missing something.

**Reply by yoavniran** [reply]
Looks like missing info on the documentation (and typing) side - Will be fixed. 
`useBatchAddListener` does receive both the batch and the [upload options](https://react-uploady.org/docs/api/types/#uploadoptions) object. Therefore, the param(s) added by the "Typed" field are accessible

**Reply by ardian-c** [reply]
I followed the example and this is what I have for the dropzone element: 
```
// the component
// more code from the example here ... 
<UploadDropZone
      id={id}
      className={className}
      onDragOverClassName={onDragOverClassName}
      inputFieldName={inputName}
      fileFilter={(file) => onFileFilter(file)}
      params={{ ...params, uploadType: type }}
    >
      <img src={cloudUploadIcon} alt='upload indicator' />
      <div className='d-flex flex-row align-items-center'>
        <button
          className='text-primary-700 p-0 bg-transparent border-0 h-100'
          onClick={() =>
            showFileUpload({
              clearPendingOnAdd: true,
              maxGroupSize: 1,
              fileFilter: onFileFilter,
            })
          }
        >
          Click to upload
        </button>
        &nbsp;
        <p className='m-0'>or drag and drop</p>
      </div>
      <p className='text-gray-500 mb-0'>{helperText}</p>
    </UploadDropZone>
   
 //  a use case
 const CustomDocumentUploadField = createUploadFieldType('DOCUMENT');
 <CustomDocumentUploadField
      key='document'
      id='document-upload-dropzone'
      className='custom__upload-modal-dropzone'
      onDragOverClassName='custom__upload-modal-dropzone--dragover'
/>
```
However, the `options.params` does not receive the type from here.

**Reply by yoavniran** [reply]
Hey @ardian-c 
thats strange
It will be really useful if you can put your example in a reproducing sandbox so I can take a look

**Reply by ardian-c** [reply]
@yoavniran Here is a sample [sandbox](https://codesandbox.io/s/cocky-curie-3bdc3e?file=/src/App.js)

**Reply by yoavniran** [reply]
works like a charm :)


https://user-images.githubusercontent.com/1102278/208123652-64b28730-f46c-4463-9d34-472e0a25235b.mp4

Whats throwing you off most likely is that the upload button isnt sending the param. Thats because you're only passing the `uploadType` param to the UploadDropZone. 

You need to pass the same params to your custom button (calling `showFileUpload`).

**Reply by ardian-c** [reply]
@yoavniran Great, I was definitively missing that, I did upload using only the button and that was the confusion. Thank you for your help, appreciated! 

One more thing, is it preferred to have a separate Uploady context for these components, or can they be within one context using the current solution to differentiate between them? Are there any drawbacks to these approaches?

**Reply by yoavniran** [reply]
From my experience using a single context is easier to manage and requires less duplication of code and also less memory footprint (though thats not a major concern typically).

I dont think there's a real drawback to using multiple contexts if that makes sense for your app and the way you structure it.
